### PR TITLE
backend/fix : Fixed CheckOverlap and flowStatus and sorted list sched…

### PR DESCRIPTION
**Pull request by dushyant-2002**
…uled booking

## Type of Change

- [x] Bugfix
- [ ] New feature
- [x] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

### Additional Changes

- [ ] This PR modifies the database schema (database migration added)
- [ ] This PR modifies dhall configs/environment variables
- [x] This PR contains API breaking changes

## Motivation and Context

1. fixed flow status for scheduled rides
2. sorted the list scheduled booking and only reachable bookings are visible to driver.

## Screenshot of test results

Screenshot 2024-08-28 at 5 07 48 PM

## Checklist

- [ ] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [ ] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added integration tests for my changes where possible
- [x] I tested the changes using integration tests
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable
- [ ] No leak detected in leakcanary

### CODE REVIEW

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Beckn/OnUpdate.hs [hunk 1]
```diff
@@ -400,11 +400,12 @@ onUpdate = \case
         else return False
     let newQuote = quote{id = Id quoteId_, createdAt = now, updatedAt = now}
         newBooking = booking{id = bookingId, quoteId = Just (Id quoteId_), status = SRB.CONFIRMED, isScheduled = newIsScheduled, bppBookingId = Just newBppBookingId, startTime = max now booking.startTime, createdAt = now, updatedAt = now}
+        flowStatus = if newIsScheduled then DPFS.IDLE else DPFS.WAITING_FOR_DRIVER_ASSIGNMENT {bookingId = bookingId, validTill = searchReq.validTill, fareProductType = Just $ STB.getFareProductType booking.bookingDetails, tripCategory = booking.tripCategory}
```
Comment (0utkarsh) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Beckn/OnUpdate.hs:403]: Why this change?

File: Backend/app/rider-platform/rider-app/Main/src/SharedLogic/Confirm.hs [hunk 2]
```diff
@@ -176,31 +172,58 @@ confirm DConfirmReq {..} = do
...
-                { origin = currBookingDest,
-                  destination = scheduledPickup,
-                  travelMode = Just Maps.CAR,
-                  distanceUnit = Meter
-                }
-          return $ distance.distance.getMeters
+isRidePossible :: UTCTime -> UTCTime -> UTCTime -> UTCTime -> Bool
+isRidePossible bookingStartTime bookingEndTime searchStartTime searchEndTime =
+  (bookingStartTime <= searchStartTime && searchStartTime <= bookingEndTime)
+    || (bookingStartTime <= searchEndTime && searchEndTime <= bookingEndTime)
```
Comment (0utkarsh) [inline Backend/app/rider-platform/rider-app/Main/src/SharedLogic/Confirm.hs:185]: `searchStartTime > bookingEndTime || searchEndTime < bookingStartTime` - Only this should be enough?

File: Backend/app/rider-platform/rider-app/Main/src/SharedLogic/Confirm.hs [hunk 3]
```diff
@@ -176,31 +172,58 @@ confirm DConfirmReq {..} = do
...
+      case searchReq.roundTrip of
+        Just True ->
+          let endTime = fromMaybe searchReq.startTime searchReq.returnTime
+           in addUTCTime merchant.scheduleRideBufferTime endTime
+        _ ->
+          let estimatedDistanceInKm = calculateEstimatedDistanceInKm searchReq.distance
+           in addUTCTime (intToNominalDiffTime (estimatedDistanceInKm * 3 * 60) + merchant.scheduleRideBufferTime) searchReq.startTime
+    _ ->
+      let estimatedDistanceInKm = calculateEstimatedDistanceInKm searchReq.distance
+       in addUTCTime (intToNominalDiffTime (estimatedDistanceInKm * 3 * 60) + merchant.scheduleRideBufferTime) searchReq.startTime
```
Comment (0utkarsh) [inline Backend/app/rider-platform/rider-app/Main/src/SharedLogic/Confirm.hs:221]: Logic of both functions is similar, can you refactor?

**Review by 0utkarsh — CHANGES_REQUESTED** [review]

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Beckn/OnUpdate.hs (hunk 1 above)
Comment (dushyant-2002) [reply Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Beckn/OnUpdate.hs:403]: when driver accepts booking lets say before 2 hours then he cancels it then the screeen was going to reallocation screen but we dont want that screen . so here newIsScheduled will be true in that case if cancelled before more than 30 minutes and in case of false will show reallocation screen

**Review by 0utkarsh — APPROVED** [review]
